src/BinSkim.Driver/RoslynAnalyzer/ILImporter.cs

```diff
@@ -0,0 +1,741 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Reflection.Metadata;
+
+using Microsoft.CodeAnalysis.Semantics;
+
+using ILExceptionRegion = System.Reflection.Metadata.ExceptionRegion;
+
+namespace Microsoft.CodeAnalysis.IL
+{
+    internal partial class ILImporter
+    {
+        private Compilation _compilation;
+        private IMethodSymbol _method;
+        private byte[] _ilBytes;
+        private ExceptionRegion[] _exceptionRegions;
+        private List<IStatement> _statements;
+        private int _labelCount;
+
+        public ILImporter(Compilation compilation, IMethodSymbol method, MethodBodyBlock body)
+        {
+            if (body.ExceptionRegions.Length != 0)
+            {
+                throw new NotImplementedException();
+            }
+
+            _compilation = compilation;
+            _method = method;
+            _ilBytes = body.GetILBytes();
+            _exceptionRegions = new ExceptionRegion[0];
+            _statements = new List<IStatement>();
+        }
+
+        public IBlockStatement Import()
+        {
+            FindBasicBlocks();
+            ImportBasicBlocks();
+
+            var statements = ImmutableArray.CreateBuilder<IStatement>(_statements.Count + _labelCount);
+
+            foreach (var block in _basicBlocks)
+            {
+                if (block != null)
+                {
+                    if (block.Label != null)
+                    {
+                        statements.Add(new LabelStatement(block.Label));
+                    }
+
+                    for (int i = block.FirstStatementOffset; i <= block.LastStatementOffset; i++)
+                    {
+                        statements.Add(_statements[i]);
+                    }
+                }
+            }
+
+            return new BlockStatement(ImmutableArray<ILocalSymbol>.Empty, statements.MoveToImmutable());
+        }
+
+        private void Push(StackValue value)
+        {
+            if (_stackTop >= _stack.Length)
+            {
+                Array.Resize(ref _stack, 2 * _stackTop + 3);
```

Comment:
If memory use is your concern maybe using 1.5 instead of 2 would work (this is what the STL's std::vector does) but switching to fixed-size allocations would cause algorithms to degrade to quadradic time at that threshold (which is exactly where you don't want them degrading to O(n^2) :) )